### PR TITLE
GridFunction::CurlDim() nullptr fix

### PR DESCRIPTION
**Pull request by psocratis**
| PR | Author | Editor | Reviewers | Assignment | Approval | Merge|
 | --- | --- | --- | --- | --- | --- | --- |
| [#4330](https://github.com/mfem/mfem/pull/4330) | @psocratis | @tzanio | @mlstowell + @dylan-copeland | 5/29/24 | 6/1/24 | 6/5/24 | |

<details>
<summary>PR Checklist</summary>
    
- [ ] Code builds.
- [ ] Code passes `make style`.
- [ ] Update `CHANGELOG`:
    - [ ] Is this a new feature users need to be aware of? New or updated example or miniapp?
    - [ ] Does it make sense to create a new section in the `CHANGELOG` to group with other related features?
- [ ] Update `INSTALL`:
    - [ ] Had a new optional library been added? If so, what range of versions of this library are required? (*Make sure the external library is compatible with our BSD license, e.g. it is not licensed under GPL!*)
    - [ ] Have the version ranges for any required or optional libraries changed?
    - [ ] Does `make` or `cmake` have a new target?
    - [ ] Did the requirements or the installation process change? *(rare)*
- [ ] Update continuous integration server configurations if necessary (e.g. with new version requirements for each of MFEM's dependencies)
    - [ ] `.github`
    - [ ] `.appveyor.yml`
- [ ] Update `.gitignore`:
    - [ ] Check if `make distclean; git status` shows any files that were generated from the source by the project (not an IDE) but we don't want to track in the repository.
    - [ ] Add new patterns (just for the new files above) and re-run the above test.
- [ ] New examples:
    - [ ] All sample runs at the top of the example source file work.
    - [ ] Update `examples/makefile`:
      - [ ] Add the example code to the appropriate `SEQ_EXAMPLES` and `PAR_EXAMPLES` variables.
      - [ ] Add any files generated by it to the `clean` target.
      - [ ] Add the example binary and any files generated by it to the top-level `.gitignore` file.
    - [ ] Update `examples/CMakeLists.txt`:
      - [ ] Add the example code to the `ALL_EXE_SRCS` variable.
      - [ ] Make sure `THIS_TEST_OPTIONS` is set correctly for the new example.
   - [ ] List the new example in `doc/CodeDocumentation.dox`.
   - [ ] If new examples directory (e.g.`examples/pumi`), list it in `doc/CodeDocumentation.conf.in`
   - [ ] Companion pull request for documentation in [mfem/web](https://github.com/mfem/web) repo:
      - [ ] Update or add example-specific documentation, see e.g. the `src/examples.md`.
      - [ ] Add the description, labels and screenshots in `src/examples.md` and `src/img`.
      - [ ] In `examples.md`, list the example under the appropriate categories, add new categories if necessary.
      - [ ] Add a short description of the example in the "Extensive Examples" section of `features.md`.
- [ ] New miniapps:
   - [ ] All sample runs at the top of the miniapp source file work.
   - [ ] Update top-level `makefile` and `makefile` in corresponding miniapp directory.
   - [ ] Add the miniapp binary and any files generated by it to the top-level `.gitignore` file.
   - [ ] Update CMake build system:
      - [ ] Update the `CMakeLists.txt` file in the `miniapps` directory, if the new miniapp is in a new directory.
      - [ ] Add/update the `CMakeLists.txt` file in the new miniapp directory.
      - [ ] Consider adding a new test for the new miniapp.
   - [ ] List the new miniapp in `doc/CodeDocumentation.dox`
   - [ ] If new miniapps directory (e.g.`miniapps/nurbs`), add it to `MINIAPP_SUBDIRS` in the `makefile`.
   - [ ] If new miniapps directory (e.g.`miniapps/nurbs`), list it in `doc/CodeDocumentation.conf.in`
   - [ ] Companion pull request for documentation in [mfem/web](https://github.com/mfem/web) repo:
     - [ ] Update or add miniapp-specific documentation, see e.g. the `src/meshing.md` and `src/electromagnetics.md` files.
     - [ ] Add the description, labels and screenshots in `src/examples.md` and `src/img`.
     - [ ] The miniapps go at the end of the page, and are usually listed only under a specific "Application (PDE)" category.
     - [ ] Add a short description of the miniapp in the "Extensive Examples" section of `features.md`.
- [ ] New capability:
   - [ ] All new public, protected, and private classes, methods, data members, and functions have full Doxygen-style documentation in source comments. Documentation should include descriptions of member data, function arguments and return values, template parameters, and prerequisites for calling new functions.
   - [ ] Pointer arguments and return values must specify whether ownership is being transferred or lent with the call.
   - [ ] Any new functions should include descriptions of their intended use e.g. for internal use only, user-facing, etc., along with references to example code whenever possible/appropriate.
   - [ ] Consider adding new sample runs in existing examples to highlight the new capability.
   - [ ] Consider saving cool simulation pictures with the new capability in the Confluence gallery (LLNL only) or submitting them, via pull request, to the gallery section of the `mfem/web` repo.
   - [ ] If this is a major new feature, consider mentioning it in the short summary inside `README` *(rare)*.
   - [ ] List major new classes in `doc/CodeDocumentation.dox` *(rare)*.
- [ ] Update this checklist, if the new pull request affects it.
- [ ] Run `make unittest` to make sure all unit tests pass.
- [ ] Run the tests in `tests/scripts`.
- [ ] (LLNL only) After merging:
   - [ ] Update internal tests to include the new features.
</details>

### CODE REVIEW

**Review by mlstowell — APPROVED** [review]
Thanks for catching and fixing this, @psocratis

**Comment by tzanio** [comment]
This PR is now [under review](https://github.com/mfem/mfem/blob/master/CONTRIBUTING.md#pull-requests) (see the table in the PR description). To help with the review process, please **do not force push** to the branch.

File: fem/gridfunc.cpp [hunk 1]
```diff
@@ -348,9 +348,11 @@ int GridFunction::CurlDim() const
    const FiniteElement *fe;
    if (!fes->GetNE())
    {
+      const FiniteElementCollection *fe_coll = fes->FEColl();
```
Comment (dylan-copeland) [inline fem/gridfunc.cpp:351]: Since this is just used once, it doesn't really help to define `fe_coll`.

**Review by dylan-copeland — APPROVED** [review]

**Comment by tzanio** [comment]
Merged in `next` for testing...

**Comment by tzanio** [comment]
Re-merged in `next` for testing...

**Comment by tzanio** [comment]
Unfortunatelly df09aea4daac122d9b79b6619ceada2d2adc7a12 is a breaking change as some code did use `fec` directly.

We have an autotest failure because of that in `mesh/pumi.cpp:893`.

There may be other codes that use this, we should either revert it or mention the change explicitly in `CHANGELOG`, @v-dobrev what do you think?

**Comment by v-dobrev** [comment]
> Unfortunatelly [df09aea](https://github.com/mfem/mfem/commit/df09aea4daac122d9b79b6619ceada2d2adc7a12) is a breaking change as some code did use `fec` directly.
> 
> We have an autotest failure because of that in `mesh/pumi.cpp:893`.
> 
> There may be other codes that use this, we should either revert it or mention the change explicitly in `CHANGELOG`, @v-dobrev what do you think?

I'm not sure the rename `fec` -> `fec_owned` makes things clearer (if that was the motivation for it) -- the doxygen should be sufficient. If I had to choose, I'd revert the rename, just to be backward compatible. However, I'm not strongly opposed to keeping it with the new name as long as it is documented in CHANGELOG.

I'm not sure if there's a way to keep `fec` as deprecated (so codes that use it will get a deprecation warning) and have the new name too. We do something like that with `FaceElementTransformations::FaceGeom` which is a deprecated reference to another data member -- maybe we can make `fec` a deprecated reference to `fec_owned`.

**Comment by dylan-copeland** [comment]
The motivation for renaming `fec` -> `fec_owned` is that MFEM commonly (in many places) uses `fes` and `fec` together, but here `fec` has a special meaning. Following the habit of using `fec` like usual can cause bugs. This PR fixed a bug where `fes->FEColl()` should be used instead of `fec`. These are different and can be easily confused. Making a clearly different name is an improvement that can prevent bugs in this class and derived classes. The disadvantage is the breaking change that affects derived classes like in the pumi code, but to me it seems worth the small change to prevent bugs.

**Comment by v-dobrev** [comment]
I pushed a fix for the nightly test failure. For backward compatibility, I also added `GridFunction::fec` as a deprecated reference to `GridFunction::fec_owned` -- if codes use `fec` they will get deprecation warnings but still work.

**Comment by v-dobrev** [comment]
Hmm, GCC does not like the initialization of the deprecated field. I tested just with clang which is ok. Let me try to find a workaround.

**Comment by v-dobrev** [comment]
> Hmm, GCC does not like the initialization of the deprecated field. I tested just with clang which is ok. Let me try to find a workaround.

I can't find a workaround for GCC. I guess we can either:

1. remove `MFEM_DEPRECATED` from the declaration of `fec` -- there will be no compiler warnings for using `fec`, just the doxygen deprecation comment, or
2. remove `fec` altogather, as it was originally in this PR -- users will get an error if using `fec`.

What do you prefer, @psocratis, @tzanio, @dylan-copeland, @mlstowell?

In either case, it will be good to mention the deprecation/removal of `GridFunction::fec` in the CHANGELOG.

**Comment by psocratis** [comment]
> What do you prefer, @psocratis, @tzanio, @dylan-copeland, @mlstowell?

I would go for option 2

**Comment by v-dobrev** [comment]
Re-merged in `next` for testing...

**Comment by tzanio** [comment]
I also like option 2 with a mention in `CHANGELOG`

**Review by tzanio — APPROVED** [review]

**Pull request merged**